client/notices.go

```diff
@@ -0,0 +1,189 @@
+		return "", err
+	}
+	return result.ID, err
+}
+
+type NoticesOptions struct {
+	// Type, if set, includes only notices of this type.
+	Type NoticeType
+	// Key, if set, includes only notices with this key.
+	Key string
+	// After, if set, includes only notices that were last repeated after this time.
+	After time.Time
+}
+
+// Notice is a notification whose identity is the combination of Type and Key
+// that has occurred Occurrences number of times.
+type Notice struct {
+	ID            string            `json:"id"`
+	Type          NoticeType        `json:"type"`
+	Key           string            `json:"key"`
+	FirstOccurred time.Time         `json:"first-occurred"`
+	LastOccurred  time.Time         `json:"last-occurred"`
+	LastRepeated  time.Time         `json:"last-repeated"`
+	Occurrences   int               `json:"occurrences"`
+	LastData      map[string]string `json:"last-data,omitempty"`
+	RepeatAfter   time.Duration     `json:"repeat-after,omitempty"`
+	ExpireAfter   time.Duration     `json:"expire-after,omitempty"`
+}
+
+type NoticeType string
+
+const (
+	// Recorded whenever a change is updated: when it is first spawned or its
+	// status was updated.
+	NoticeChangeUpdate NoticeType = "change-update"
+
+	// A client notice reported via the Pebble client API or "pebble notify".
+	// The key and data fields are provided by the user. The key must be in
+	// the format "mydomain.io/mykey" to ensure well-namespaced notice keys.
+	NoticeClient NoticeType = "client"
+
+	// Warnings are a subset of notices where the key is a human-readable
+	// warning message.
+	NoticeWarning NoticeType = "warning"
+)
+
+type jsonNotice struct {
+	Notice
+	RepeatAfter string `json:"repeat-after,omitempty"`
+	ExpireAfter string `json:"expire-after,omitempty"`
+}
+
+// Notice fetches a single notice by ID.
+func (client *Client) Notice(id string) (*Notice, error) {
+	var jn *jsonNotice
+	_, err := client.doSync("GET", "/v1/notices/"+id, nil, nil, nil, &jn)
+	if err != nil {
+		return nil, err
+	}
+	return jsonNoticeToNotice(jn), nil
+}
+
+// Notices returns a list of notices that match the filters given in opts,
+// ordered by the last-repeated time.
+func (client *Client) Notices(opts *NoticesOptions) ([]*Notice, error) {
+	query := makeNoticesQuery(opts)
+	var jns []*jsonNotice
+	_, err := client.doSync("GET", "/v1/notices", query, nil, nil, &jns)
+	return jsonNoticesToNotices(jns), err
+}
+
+// WaitNotices returns a list of notices that match the filters given in opts,
+// waiting up to the given timeout. They are ordered by the last-repeated time.
+func (client *Client) WaitNotices(ctx context.Context, opts *NoticesOptions, timeout time.Duration) ([]*Notice, error) {
+	query := makeNoticesQuery(opts)
+	query.Set("timeout", timeout.String())
+
+	res, err := client.raw(ctx, "GET", "/v1/notices", query, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode == http.StatusGatewayTimeout {
+		return nil, nil
```

Comment:
shouldn't we surface this error?

Reply:
This is intentional, though I've now clarified that in the doc comment. I think it makes sense for this to not be an error, because this is intended as a "long-polling" style API, and the timeout you asked for just elapsed (not an error condition). It makes it easy for callers who are long polling to just check `len(notices) == 0` and try again.

Reply:
Ok interesting I see. 

So this is the reason why you had to split the ```doSync```? 

I made a comment on this topic here in the [API PR](https://github.com/canonical/pebble/pull/296/files#r1324706774). I completely second the fact that this should not error. My question would just be what is the point of sending back an error in the first place, as apposed to an empty notices list (like for this case where ```--timeout``` is not supplied).

Reply:
No, the reason I needed to split `doSync` was to use `raw()` in `WaitNotices` so I could pass the context through. I've added a comment to that effect.